primitives/consensus/beefy/src/lib.rs

```diff
@@ -302,6 +337,56 @@ where
 	return valid_first && valid_second
 }
 
+/// Validates [ForkEquivocationProof] by checking:
+/// 1. `commitment` is signed,
+/// 2. `correct_header` is valid and matches `commitment.block_number`.
+/// 2. `commitment.payload` != `expected_payload(correct_header)`.
+/// NOTE: GRANDPA finalization proof is not checked, which leads to slashing on forks.
+/// This is fine since honest validators will not be slashed on the chain finalized
+/// by GRANDPA, which is the only chain that ultimately matters.
+/// The only material difference not checking GRANDPA proofs makes is that validators
+/// are not slashed for signing BEEFY commitments prior to the blocks committed to being
+/// finalized by GRANDPA. This is fine too, since the slashing risk of committing to
+/// an incorrect block implies validators will only sign blocks they *know* will be
+/// finalized by GRANDPA.
+pub fn check_fork_equivocation_proof<Number, Id, MsgHash, Header>(
+	proof: &ForkEquivocationProof<Number, Id, <Id as RuntimeAppPublic>::Signature, Header>,
+	expected_header_hash: &Header::Hash,
+) -> bool
+where
+	Id: BeefyAuthorityId<MsgHash> + PartialEq,
+	Number: Clone + Encode + PartialEq,
+	MsgHash: Hash,
+	Header: sp_api::HeaderT,
+{
+	let ForkEquivocationProof { commitment, signatories, correct_header } = proof;
+
+	if correct_header.hash() != *expected_header_hash {
+		return false
+	}
+
+	let expected_mmr_root_digest = mmr::find_mmr_root_digest::<Header>(correct_header);
+	let expected_payload = expected_mmr_root_digest.map(|mmr_root| {
+		Payload::from_single_entry(known_payloads::MMR_ROOT_ID, mmr_root.encode())
+	});
+
+	// cheap failfasts:
+	// 1. check that `payload` on the `vote` is different that the `expected_payload`
+	// 2. if the signatories signed a payload when there should be none (for
+	// instance for a block prior to BEEFY activation), they should likewise be
+	// slashed
+	if Some(&commitment.payload) != expected_payload.as_ref() || expected_mmr_root_digest.is_none() {
```

Comment:
I think the first check in enough - it already matches for expected_payload.is_none(), right?

```suggestion
	if Some(&commitment.payload) != expected_payload.as_ref() {
```